src/Orleans/Providers/ILeaseProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.LeaseProviders
+{
+    /// <summary>
+    /// Acquired lease
+    /// </summary>
+    public class AcquiredLease 
+    {
+        /// <summary>
+        /// The resource key which the lease is attached to 
+        /// </summary>
+        public string ResourceKey { get; }
+        /// <summary>
+        /// Duration of the aquired lease
+        /// </summary>
+        public TimeSpan Duration { get; }
+        /// <summary>
+        /// Lease token, whcih will be null if acquiring or renewing the lease failed
+        /// </summary>
+        public string Token { get; }
+        /// <summary>
+        /// Start time for this lease, which is when the lease is acquired or renewed
+        /// </summary>
+        public DateTime StartTimeUtc { get; }
+    }
+
+    /// <summary>
+    /// AcquireLeaseResult class, which demonstrates result of acquiring or renewing lease operation
+    /// </summary>
+    public class AcquireLeaseResult
+    {
+        /// <summary>
+        /// Acquired lease, which will be null if acquire or renew operation failed.
+        /// </summary>
+        AcquiredLease AcquiredLease { get; }
+        /// <summary>
+        /// Whether successfully acquire or renew the lease
+        /// </summary>
+        public bool Success { get; }
+        /// <summary>
+        /// If acquiring or renewing the lease failed, this is the exception which caused it. This field would be null if operation succeed.
+        /// </summary>
+        public Exception FailureException { get; }
```

Comment:
After talking with Reuben, we are more in an agreement on this 

 ```
/// <summary>
    /// AcquireLeaseResult class, which demonstrates result of acquiring or renewing lease operation
    /// </summary>
    public class AcquireLeaseResult
    {
        /// <summary>
        /// Acquired lease, which will be null if acquire or renew operation failed.
        /// </summary>
        AcquiredLease AcquiredLease { get; }
        /// <summary>
        /// Response code
        /// </summary>
        public ResponseCode StatusCode { get; }
        /// <summary>
        /// If acquiring or renewing the lease failed, this is the exception which caused it. This field would be null if operation succeed.
        /// </summary>
        public Exception FailureException { get; }
    }

    public enum ResponseCode
    {
        /// <summary>
        /// Lease is owned by other entity
        /// </summary>
        LeaseNotAvailable,
        /// <summary>
        /// The token in the AcquiredLease is invalid, which means the lease expired
        /// </summary>
        InValidToken,
        /// <summary>
        /// Operation succeed
        /// </summary>
        Succeed,
        /// <summary>
        /// TransientFailure, which should be retriable. 
        /// </summary>
        TransientFailure
    }
```
StatusCode will be the main point which the caller side decide its flow control on. But still keeping the FailureException field, as an extensibility point to provide more information for diagnose. For example, if user build the LeaseProvider on top of azure blob, then we can pass in native azure blob exception in the FailureException field.